### PR TITLE
Fix tracking of finished jobs within JobMaster

### LINKED ISSUE #9874: Failed to load the data due to the job service is busy
**Alluxio Version:**
2.1.0-snapshot

**Describe the bug**

**To Reproduce**

1. Set replication

```
/opt/alluxio/bin/alluxio fs setReplication --max 1 --min 0 /oss-gw/images
``` 

2. load the data 

```
time /opt/alluxio/bin/alluxio fs -Dalluxio.user.ufs.block.read.location.policy=alluxio.client.block.policy.DeterministicHashPolicy  distributedLoad --replication 1 /oss-gw/images/
Failed to successfully complete the job.
```

3. Check the logs

```
kubectl logs -f alluxio-master-0 -f alluxio-job-master
2019-09-11 08:45:57,091 INFO  JobCoordinator - Starting job EvictConfig{blockId=3271557122, replicas=1}
2019-09-11 08:45:57,093 WARN  JobMasterClientServiceHandler - Exit (Error): run: request=jobConfig: "\254\355\000\005sr\000!alluxio.job.replicate.EvictConfig\f\353\231\346 \2303\311\002\000\002J\000\bmBlockIdI\000\tmReplicasxp\000\000\000\000\235\000\000\000\000\000\000\001"
, Error=alluxio.exception.status.ResourceExhaustedException: Job master is at full capacity of 100,000 jobs
```

```
kubectl logs -f alluxio-master-0 -f alluxio-master
2019-09-11 08:42:44,031 WARN  SleepingTimer - Master Replication Check last execution took 64010 ms. Longer than the interval 60000
2019-09-11 08:43:48,044 WARN  ReplicationChecker - The job service is busy, will retry later. alluxio.exception.status.ResourceExhaustedException: Job master is at full capacity of 100,000 jobs
2019-09-11 08:43:48,044 WARN  SleepingTimer - Master Replication Check last execution took 64012 ms. Longer than the interval 60000
2019-09-11 08:44:52,057 WARN  ReplicationChecker - The job service is busy, will retry later. alluxio.exception.status.ResourceExhaustedException: Job master is at full capacity of 100,000 jobs
2019-09-11 08:44:52,058 WARN  SleepingTimer - Master Replication Check last execution took 64014 ms. Longer than the interval 60000
2019-09-11 08:45:56,078 WARN  ReplicationChecker - The job service is busy, will retry later. alluxio.exception.status.ResourceExhaustedException: Job master is at full capacity of 100,000 jobs
2019-09-11 08:45:56,078 WARN  SleepingTimer - Master Replication Check last execution took 64020 ms. Longer than the interval 60000
2019-09-11 08:45:57,093 WARN  ReplicationChecker - The job service is busy, will retry later. alluxio.exception.status.ResourceExhaustedException: Job master is at full capacity of 100,000 jobs
```

4. But it's strange that some data has been loaded.

```
/opt/alluxio/bin/alluxio fsadmin report capacity
Capacity information for all workers:
    Total Capacity: 600.00GB
        Tier: MEM  Size: 200.00GB
        Tier: SSD  Size: 400.00GB
    Used Capacity: 112.02GB
        Tier: MEM  Size: 112.02GB
        Tier: SSD  Size: 0B
    Used Percentage: 18%
    Free Percentage: 82%

Worker Name      Last Heartbeat   Storage       Total            MEM           SSD
192.168.0.194    0                capacity      150.00GB         50.00GB       100.00GB
                                  used          27.72GB (18%)    27.72GB       0B
192.168.0.117    0                capacity      150.00GB         50.00GB       100.00GB
                                  used          29.46GB (19%)    29.46GB       0B
192.168.0.118    0                capacity      150.00GB         50.00GB       100.00GB
                                  used          26.71GB (17%)    26.71GB       0B
192.168.0.195    0                capacity      150.00GB         50.00GB       100.00GB
                                  used          28.13GB (18%)    28.13GB       0B
```

**Expected behavior**

**Urgency**
Describe the impact and urgency of the bug.

**Additional context**

### PR DESCRIPTION
**Pull request by ZacBlanco**
After some investigation it was found that the JobMaster was not properly
tracking finished jobs due to discrepancies in the implementation of the 
`compareTo` and `equals` methods of `JobInfo`. This led to jobs being
removed from the ConcurrentSkipListSet without ever actually calling
the `remove` function. The resulting behavior is that we would see the
master hit the maximum job capacity very frequently as there would be
un-flushable jobs in the `mIdToCoordinators` map.

This change fixes the original issue by altering the underlying structure
used to keep track of the finished jobs and by making more guarantees
about how the status of jobs may change.

First, with this change we no longer allow a JobInfo to change states after
it has been marked with a state where `JobInfo#isFinished` returns true.
This means all jobs states should now follow a DAG (RUNNING is an
optional state as a job may move from CREATED to any one of the finished
states):

```
CREATED -> (RUNNING) -> [FAILED|CANCELLED|COMPLETED]
```

Second, instead of utilizing a ConcurrentSkipListSet I've opted for a
simpler concurrent structure; the LinkedBlockingQueue. Using a FIFO
queue with the guarantee that once jobs are finished they cannot change
states, we gain the opportunity to use a simple queue that has O(1)
offer/poll operations. With this approach jobs that are marked as finished
may get added in different orders depending on how the internal locks
within the queue operate. This implementation no longer strictly guarantees
that objects in the queue are ordered by the lastStatusChangeMs time. The
amount of time difference between concurrent jobs being added should not
be so great that leaves jobs such that the retention time severely affects
the amount of jobs that can be evicted.

This implementation introduces the JobTracker class which is now
responsible for adding and evicting jobs from the job master. It encapsulates
what was the mIdToCoordinator map, and also houses the FIFO queue that
used to be the mFinishedJobs set.

Closes #9874

### CODE REVIEW

**Comment by ZacBlanco** [comment]
I had done some original investigation trying to root-cause the issue in #9932 but couldn't figure out exactly what was causing the "ghosting" from the `mFinishedJobs` map.

I took a new approach to tracking jobs through a priority queue instead of the ConcurrentSkipListSet. The PR is still WIP as it needs to be cleaned up quite a bit. For now, everything seems to be fixed.

File: job/server/src/main/java/alluxio/master/job/JobTracker.java [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+   * One caveat to setting this value is that there will be a lower bound on the amount of
+   * memory that the job master will use when it is at capacity. This may or may not be a
+   * reasonable trade-off to guarantee that RPCs don't time out due to job eviction.
+   * @see PropertyKey#JOB_MASTER_FINISHED_JOB_PURGE_COUNT
+   */
+  private final long mMaxJobPurgeCount;
+
+  /**
+   * The maximum amount of jobs that can be tracked at any one time.
+   */
```
Comment (calvinjia) [inline job/server/src/main/java/alluxio/master/job/JobTracker.java]: Do you mind using `/** comment */`?

File: job/server/src/main/java/alluxio/master/job/JobTracker.java [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+    mJobIdGenerator = new JobIdGenerator();
+  }
+
+  private void statusChangeCallback(JobInfo jobInfo) {
+    if (jobInfo == null) {
+      return;
+    }
+    Status status = jobInfo.getStatus();
+    if (status.isFinished()) {
+      if (mFinished.offer(jobInfo)) {
```
Comment (calvinjia) [inline job/server/src/main/java/alluxio/master/job/JobTracker.java]: Missing `!`?

File: job/server/src/main/java/alluxio/master/job/JobTracker.java [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+        LOG.warn("Failed to add job to finished queue");
+        // TODO(zac) If adding to the finished queue fails, should we wait and retry?
+        // another option is to at this point just immediately remove from tracking so we don't
+        // end up with ghost jobs, or provide some way through the REST API to remove jobs from
+        // mCoordinators at runtime
```
Comment (calvinjia) [inline job/server/src/main/java/alluxio/master/job/JobTracker.java]: Perhaps evict old jobs?

File: job/server/src/main/java/alluxio/master/job/JobTracker.java [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+    } else {
+      throw new ResourceExhaustedException(
+          ExceptionMessage.JOB_MASTER_FULL_CAPACITY.getMessage(mCapacity));
+    }
+  }
+
+  /**
+   * Removes all finished jobs outside of the retention time from the queue.
+   *
+   * @return true if at least one job was removed, or if the initial
```
Comment (calvinjia) [inline job/server/src/main/java/alluxio/master/job/JobTracker.java]: complete comment?

File: job/server/src/main/java/alluxio/master/job/JobTracker.java [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+   */
+  private synchronized boolean removeFinished() {
+    boolean removedJob = false;
+    boolean isFull = mCoordinators.size() >= mCapacity;
+    if (isFull) { // coordinators at max capacity
```
Comment (calvinjia) [inline job/server/src/main/java/alluxio/master/job/JobTracker.java]: One coding style we try to follow is to put the early returns first - ie.
```
if (!isFull) {
  return true;
}
// isFull logic here
```

File: job/server/src/main/java/alluxio/master/job/JobTracker.java (hunk 3 above)
Comment (ZacBlanco) [reply job/server/src/main/java/alluxio/master/job/JobTracker.java]: See 3b101a0 for my approach

**Comment by ZacBlanco** [comment]
Additionally, I would recommend this change be ported into `branch-2.0`

File: job/server/src/main/java/alluxio/master/job/JobCoordinator.java [hunk 6]
```diff
@@ -190,9 +204,13 @@ public synchronized void failTasksForWorker(long workerId) {
     if (taskInfo.getStatus().isFinished()) {
       return;
     }
-    taskInfo.setStatus(Status.FAILED);
-    taskInfo.setErrorMessage("Job worker was lost before the task could complete");
-    updateStatus();
+    synchronized (mJobInfo) {
```
Comment (madanadit) [inline job/server/src/main/java/alluxio/master/job/JobCoordinator.java]: same here. redundant synchronization?

**Comment by ZacBlanco** [comment]
alluxio-bot, merge this please

**Comment by ZacBlanco** [comment]
alluxio-bot, cherry pick this to branch-2.0 please

**Pull request merged**